pkg/softtainter/softtainter.go

```diff
@@ -0,0 +1,507 @@
+	appropriatelyutilizedSoftTaintValue = "true"
+)
+
+var (
+	log = logf.Log.WithName("controller_nodeClassifier")
+)
+
+type softTainterArgs struct {
+	useDeviationThresholds bool
+	thresholds             api.ResourceThresholds
+	targetThresholds       api.ResourceThresholds
+	applySoftTaints        bool
+	mode                   desv1.Mode
+}
+
+type softTainter struct {
+	args                              *softTainterArgs
+	resourceNames                     []corev1.ResourceName
+	previousPrometheusClientTransport *http.Transport
+	currentPrometheusAuthToken        string
+	promClient                        promapi.Client
+	promQuery                         string
+	promURL                           string
+	client                            client.Client
+	resyncPeriod                      time.Duration
+	policyConfigFile                  string
+	nodeUtilizationFactory            func(promapi.Client, string) NodeUtilization
+}
+
+type NodeUtilization interface {
+	NodeUsageFromPrometheusMetrics(ctx context.Context) (map[string]map[corev1.ResourceName]*resource.Quantity, error)
+}
+
+type ActualNodeUtilization struct {
+	promClient promapi.Client
+	promQuery  string
+}
+
+func (anu *ActualNodeUtilization) NodeUsageFromPrometheusMetrics(ctx context.Context) (map[string]map[corev1.ResourceName]*resource.Quantity, error) {
+	return nodeutilization.NodeUsageFromPrometheusMetrics(ctx, anu.promClient, anu.promQuery)
+}
+
+func (st *softTainter) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+
+	des := desv1.KubeDescheduler{}
+
+	descheduler.SetupPlugins()
+	policy, err := descheduler.LoadPolicyConfig(st.policyConfigFile, nil, pluginregistry.PluginRegistry)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = st.client.Get(ctx, client.ObjectKey{
+		Namespace: operatorclient.OperatorNamespace,
+		Name:      operatorclient.OperatorConfigName,
+	}, &des)
+	if err != nil {
+		logger.Error(err, "failed reading descheduler operator CR")
+		return reconcile.Result{}, err
+	}
+
+	if des.Spec.DeschedulingIntervalSeconds == nil {
+		return reconcile.Result{}, fmt.Errorf("descheduler should have an interval set")
+	}
+	st.resyncPeriod = time.Duration(*des.Spec.DeschedulingIntervalSeconds) * time.Second
+
+	var lnargs *nodeutilization.LowNodeUtilizationArgs
+
+	for _, p := range policy.Profiles {
+		if p.Name == string(desv1.LongLifecycle) || p.Name == string(desv1.RelieveAndMigrate) {
+			for _, pc := range p.PluginConfigs {
+				if pc.Name == nodeutilization.LowNodeUtilizationPluginName {
+					lnargs = pc.Args.(*nodeutilization.LowNodeUtilizationArgs)
+				}
+			}
+		}
+	}
+	if lnargs == nil {
+		err := fmt.Errorf("unable to read LowNodeUtilizationArgs")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+
+	enableSoftTainter := false
```

Comment:
`enableSoftTainter := des.Spec.ProfileCustomizations != nil && des.Spec.ProfileCustomizations.DevEnableSoftTainter`

Reply:
Assuming `DevEnableSoftTainter` means soft-taint nodes it's better to avoid rendering the controller deployment completely when `DevEnableSoftTainter` is false. Or, is there any benefit of running the controller without tainting the nodes besides removing existing taints?

Reply:
No, I think we can completely skip it to save resources

Reply:
done